src/Tools/AI Test Toolkit/src/Logs/AITLogEntry.Codeunit.al

```diff
@@ -5,21 +5,163 @@
 
 namespace System.TestTools.AITestToolkit;
 
+using System.TestTools.TestRunner;
+
 codeunit 149032 "AIT Log Entry"
 {
     Access = Internal;
 
+    var
+        LineNoFilterLbl: Label 'Codeunit %1 "%2" (Input: %3)', Locked = true;
+
     procedure DrillDownFailedAITLogEntries(AITSuiteCode: Code[100]; LineNo: Integer; VersionNo: Integer)
     var
         AITLogEntries: Record "AIT Log Entry";
+    begin
+        AITLogEntries.SetRange(Version, VersionNo);
+        DrillDownFailedAITLogEntries(AITLogEntries, AITSuiteCode, LineNo);
+    end;
+
+    procedure DrillDownFailedAITLogEntries(AITSuiteCode: Code[100]; LineNo: Integer; Tag: Text[20])
+    var
+        AITLogEntries: Record "AIT Log Entry";
+    begin
+        AITLogEntries.SetRange(Tag, Tag);
+        DrillDownFailedAITLogEntries(AITLogEntries, AITSuiteCode, LineNo);
+    end;
+
+    local procedure DrillDownFailedAITLogEntries(var AITLogEntries: Record "AIT Log Entry"; AITSuiteCode: Code[100]; LineNo: Integer)
+    var
         AITLogEntry: Page "AIT Log Entries";
     begin
         AITLogEntries.SetFilterForFailedTestProcedures();
         AITLogEntries.SetRange("Test Suite Code", AITSuiteCode);
-        AITLogEntries.SetRange(Version, VersionNo);
         if LineNo <> 0 then
             AITLogEntries.SetRange("Test Method Line No.", LineNo);
         AITLogEntry.SetTableView(AITLogEntries);
         AITLogEntry.Run();
     end;
+
+    procedure UpdateRunHistory(Code: Code[100]; LineNo: Integer; AITViewBy: Enum "AIT Run History - View By"; var TempAITRunHistory: Record "AIT Run History" temporary)
+    var
+        AITRunHistory: Record "AIT Run History";
+        SeenTags: List of [Text[20]];
+    begin
+        TempAITRunHistory.DeleteAll();
+        AITRunHistory.SetRange("Test Suite Code", Code);
```

Comment:
Are we missing any filters before AITRunHistory.FindSet()?
If not, then can we do the AITRunHistory.FindSet() once and do the check for ViewBy inside the repeat?